pkg/gds/interceptor.go

```diff
@@ -26,3 +42,71 @@ func (s *Service) serverInterceptor(ctx context.Context, in interface{}, info *g
 	log.Debug().Str("method", info.FullMethod).Str("latency", time.Since(start).String()).Err(err)
 	return out, err
 }
+
+func (s *Service) replicaInterceptor(ctx context.Context, in interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (out interface{}, err error) {
+	// Track how long the method takes to execute.
+	start := time.Now()
+
+	// Check if we're in maintenance mode - status method should still return a full response.
+	if s.conf.Maintenance && info.FullMethod != "/trisa.gds.api.v1beta1.TRISADirectory/Status" {
+		return nil, status.Error(codes.Unavailable, "the GDS service is currently in maintenance mode")
+	}
+
+	// Fetch peer information from the TLS info.
+	var peer *PeerInfo
+	if peer, err = peerFromTLS(ctx); err != nil {
+		log.Error().Err(err).Msg("unable to retrieve remote peer info")
+		// TODO: Uncomment when mTLS is added
+		//return nil, status.Error(codes.Unauthenticated, "unable to retrieve authenticated peer information")
+	}
+
+	// Add peer information to the context.
+	ctx = context.WithValue(ctx, ContextKey("peer"), peer)
+
+	// Call the handler to finalize the request and get the response.
+	out, err = handler(ctx, in)
+
+	// Log with zerolog - checkout grpclog.LoggerV2 for default logging.
+	log.Debug().Str("method", info.FullMethod).Str("latency", time.Since(start).String()).Err(err)
+	return out, err
+}
+
+// peerFromTLS looks up the TLSInfo from the incoming gRPC connection to retrieve
+// information about the remote peer from the certificate.
+func peerFromTLS(ctx context.Context) (info *PeerInfo, err error) {
+	var (
+		ok      bool
+		gp      *peer.Peer
+		tlsAuth credentials.TLSInfo
+		chains  [][]*x509.Certificate
+	)
+
+	if gp, ok = peer.FromContext(ctx); !ok {
+		return nil, errors.New("no peer found in context")
+	}
+
+	if tlsAuth, ok = gp.AuthInfo.(credentials.TLSInfo); !ok {
+		return nil, fmt.Errorf("unexpected peer transport credentials type: %T", gp.AuthInfo)
+	}
+
+	chains = tlsAuth.State.VerifiedChains
+	if len(chains) == 0 {
+		return nil, errors.New("could not find certificate chain")
+	}
+
+	// Search for a valid peer certificate.
+	for _, c := range chains {
+		// Certificate chain should contain at least the peer certificate and the CA.
+		if len(c) >= 2 {
+			// The peer certificate is always first.
+			info = &PeerInfo{
+				Name:        &c[0].Subject,
+				DNSNames:    c[0].DNSNames,
+				IPAddresses: c[0].IPAddresses,
```

Comment:
Could we perhaps also add the `Addr` from the `gp` to the start of the list of IP addresses: https://pkg.go.dev/google.golang.org/grpc/peer#Peer? 

It's possible that we will have IP addresses on the certificate, but it's usually not recommended since IP addresses can change and then you have to issue new certs. I think we should keep it here just in case it is an mTLS mechanism we use, but having the `Addr` on the list as backup would be good.